### PR TITLE
CAMS-141: Add a decision record for how to handle context

### PR DESCRIPTION
**Pull request by jamesobrooks**
# Purpose

We needed to make a [semi-]permanent decision about how to use the Azure Function Context and refactor to align with the decision.

# Major Changes

- Adds an Architectural Decision Record (ADR) to document the decision
- Removes the CaseAssignmentRepository from the ApplicationContext

# Testing/Validation

- Ran locally, ensuring that functionality is the same
- Updated automated tests

### CODE REVIEW

File: backend/functions/healthcheck/healthcheck.db.ts [hunk 1]
```diff
@@ -19,7 +19,7 @@ export default class HealthcheckCosmosDb {
   constructor(applicationContext: ApplicationContext) {
     try {
       this.ctx = applicationContext;
-      this.cosmosDbClient = getCosmosDbClient();
+      this.cosmosDbClient = getCosmosDbClient(this.ctx);
```
Comment (amorrow-flexion) [inline backend/functions/healthcheck/healthcheck.db.ts:22]: rename this to applicationContext to keep it consistent

File: backend/functions/lib/adapters/controllers/cases.controller.ts [hunk 2]
```diff
@@ -10,7 +10,7 @@ export class CasesController {
 
   constructor(context: ApplicationContext) {
     this.context = context;
-    this.caseManagement = new CaseManagement();
+    this.caseManagement = new CaseManagement(this.context);
```
Comment (amorrow-flexion) [inline backend/functions/lib/adapters/controllers/cases.controller.ts:13]: applicationContext

File: backend/functions/lib/adapters/gateways/case.assignment.cosmosdb.repository.ts [hunk 3]
```diff
@@ -18,8 +18,8 @@ export class CaseAssignmentCosmosDbRepository implements CaseAssignmentRepositor
   private cosmosConfig: CosmosConfig;
 
   constructor(context: ApplicationContext, testClient = false) {
-    this.cosmosDbClient = getCosmosDbClient(testClient);
-    this.cosmosConfig = getCosmosConfig();
+    this.cosmosDbClient = getCosmosDbClient(context, testClient);
```
Comment (amorrow-flexion) [inline backend/functions/lib/adapters/gateways/case.assignment.cosmosdb.repository.ts:21]: applicationContext

File: backend/functions/lib/adapters/utils/application-context-creator.ts [hunk 4]
```diff
@@ -9,10 +9,10 @@ export async function applicationContextCreator(
   const config = new ApplicationConfiguration();
   const featureFlags = await getFeatureFlags(config);
 
-  return {
+  const appContext = {
```
Comment (amorrow-flexion) [inline backend/functions/lib/adapters/utils/application-context-creator.ts:12]: applicationContext

File: backend/functions/lib/factory.ts [hunk 5]
```diff
@@ -15,10 +14,8 @@ export const getAttorneyGateway = (): AttorneyGatewayInterface => {
   return new AttorneyLocalGateway();
 };
 
-export const getCasesGateway = (): CasesInterface => {
-  const config: ApplicationConfiguration = new ApplicationConfiguration();
-
-  if (config.get('dbMock')) {
+export const getCasesGateway = (context: ApplicationContext): CasesInterface => {
+  if (context.config.get('dbMock')) {
```
Comment (amorrow-flexion) [inline backend/functions/lib/factory.ts:18]: applicationContext

File: backend/functions/lib/factory.ts [hunk 6]
```diff
@@ -28,28 +25,20 @@ export const getCasesGateway = (): CasesInterface => {
...
 ): CaseAssignmentRepositoryInterface => {
-  const config: ApplicationConfiguration = new ApplicationConfiguration();
-  if (config.get('dbMock')) {
-    if (context.caseAssignmentRepository) {
-      return context.caseAssignmentRepository;
-    } else {
-      context.caseAssignmentRepository = new CaseAssignmentCosmosDbRepository(context, true);
-      return context.caseAssignmentRepository;
-    }
+  if (context.config.get('dbMock')) {
```
Comment (amorrow-flexion) [inline backend/functions/lib/factory.ts:28]: applicationContext

File: backend/functions/lib/use-cases/case-management.test.ts [hunk 7]
```diff
@@ -55,7 +55,7 @@ describe('Case list tests', () => {
   });
 
   test('Calling getCases should return valid data', async () => {
-    const chapterCaseList = new CaseManagement();
+    const chapterCaseList = new CaseManagement(appContext);
```
Comment (amorrow-flexion) [inline backend/functions/lib/use-cases/case-management.test.ts:58]: applicationContext

File: backend/functions/lib/use-cases/case-management.ts [hunk 8]
```diff
@@ -12,9 +12,9 @@ import { CaseAttorneyAssignment } from '../adapters/types/case.attorney.assignme
 export class CaseManagement {
   casesGateway: CasesInterface;
 
-  constructor(casesGateway?: CasesInterface) {
+  constructor(context: ApplicationContext, casesGateway?: CasesInterface) {
```
Comment (amorrow-flexion) [inline backend/functions/lib/use-cases/case-management.ts:15]: applicationContext

**Review by amorrow-flexion — APPROVED** [review]
Built and ran the application locally to test the changes. The application functions as expected with the changes.

**Review by magile32 — APPROVED** [review]

**Pull request merged**